### PR TITLE
Create dedicated Neptune hook for metadata logging

### PR DESCRIPTION
**Pull request by AleksanderWWW**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/core/hook/neptunelogger_hook.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+@HOOKS.register_module()
+class NeptuneHook(mmvch.logger.neptune.NeptuneLoggerHook):
+    """Log metadata to Neptune.
+
+    This hook will automatically log training or evaluation metadata
+    to Neptune.ai.
+    Args:
+        interval (int): Checking interval (every k iterations).
+            Default: 50.
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: Update the args documentation to reflect all the accepted args.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    def _log_buffer(self, runner, category, log_eval=True) -> None:
+        assert category in ['epoch', 'iter']
+        # only record lr of the first param group
+        cur_lr = runner.current_lr()
+        self.base_handler['train/' + category + '/' +
+                          'learning_rate'].extend(cur_lr)
+
+        for key, value in runner.log_buffer.val_history.items():
+            self.base_handler['train/' + category + '/' + key].log(value[-1])
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: `log` -> `append`

File: mmdet/core/hook/neptunelogger_hook.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+        if log_eval and self.eval_hook._should_evaluate(runner):
+
+            results = self.eval_hook.latest_results
+
+            eval_results = self.val_dataset.evaluate(
+                results, logger='silent', **self.eval_hook.eval_kwargs)
+
+            for key, value in eval_results.items():
+                self.base_handler['val/' + category + '/' + key].log(value)
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: `log` -> `append`

File: mmdet/core/hook/neptunelogger_hook.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                runner.logger.warning(
+                    'To log evaluation NeptuneHook, '
+                    '`EvalHook` or `DistEvalHook` in mmdet '
+                    'is required, please check whether the validation '
+                    'is enabled.')
+            else:
+                self.eval_interval = self.eval_hook.interval
+                self.val_dataset = self.eval_hook.dataloader.dataset
+
+    def _log_buffer(self, runner, category, log_eval=True) -> None:
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: I think the name should be updated to reflect the fact that it also does evaluation. Wdyt?

**Review by kshitij12345 — CHANGES_REQUESTED** [review]
Thanks!

File: configs/mask_rcnn/neptune_config.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+_base_ = [
```
Comment (kshitij12345) [inline configs/mask_rcnn/neptune_config.py]: Is this file needed?

File: mmdet/core/hook/neptunelogger_hook.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from __future__ import annotations
+import os.path as osp
+from typing import Dict, Optional
+
+from mmdet.core.visualization.image import imshow_det_bboxes
+
+try:
+    import neptune.new as neptune
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: We can directly support `neptune` > 1.0 as there aren't any BC concerns.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+            Please refer to https://docs.neptune.ai/api/neptune/#init_run
+            for possible key-value pairs.
+        api_token (str): Neptune API token.
+            If None, the environment variable NEPTUNE_API_KEY will be used.
+        project (str): Neptune project name.
+            If None, the environment variable NEPTUNE_PROJECT will be used.
+        interval (int): Checking interval (every k epochs/iterations).
+            Default: 10
+        base_namespace (str): Namespace under which all the logs will
+        be present.
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: Incorrect indentation.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            Default: "training"
+        log_model (bool): If the Hook should save the final checkpoint.
+            Requires CheckpointHook.
+            Default: False
+        log_checkpoint (bool): If the Hook should log all the checkpoints
+        generated by CheckpointHook.
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: Incorrect indentation

File: mmdet/core/hook/neptunelogger_hook.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+                 interval: int = 10,
+                 base_namespace: str = 'training',
+                 log_model: bool = False,
+                 log_checkpoint: bool = False,
+                 num_eval_predictions: int = 5,
+                 **kwargs) -> None:
+        super().__init__(interval=interval, **kwargs)
+
+        if init_kwargs:
+            api_token = init_kwargs.pop('api_token', None) or api_token
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: Should we error in the case where `api_token` is passed through both `api_token` and `init_kwargs`?

File: mmdet/core/hook/neptunelogger_hook.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    def __init__(self,
+                 *,
+                 init_kwargs: Optional[Dict] = None,
+                 api_token: str = None,
+                 project: str = None,
+                 interval: int = 10,
+                 base_namespace: str = 'training',
+                 log_model: bool = False,
+                 log_checkpoint: bool = False,
+                 num_eval_predictions: int = 5,
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: Doc suggests default value is 50.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+            self.log_evaluation = False
+            runner.logger.warning(
+                'LoadImageFromFile is required to add images '
+                'to the Neptune run.')
+            return
+
+        # Select the images to be logged.
+        eval_image_indices = np.arange(len(self.val_dataset))
+        # Set seed so that same validation set is logged each time.
+        np.random.seed(42)
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: I think this is not good as it will silently set the global seed for Numpy.

We can probably reuse `self.eval_image_indices` so that we log same samples. I think it is ok to log any 50 samples between different runs.

File: configs/mask_rcnn/neptune_config.py (hunk 5 above)
Comment (AleksanderWWW) [reply configs/mask_rcnn/neptune_config.py]: Yes, it is what actually runs the hook I guess. Other integrations also have it  (e.g. https://github.com/AleksanderWWW/mmdetection/blob/master/configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_coco.py ) - though we could probably come up with a better name

File: mmdet/core/hook/neptunelogger_hook.py (hunk 9 above)
Comment (AleksanderWWW) [reply mmdet/core/hook/neptunelogger_hook.py]: Maybe let's error if they are not the same? What do you think?

File: mmdet/core/hook/neptunelogger_hook.py (hunk 11 above)
Comment (AleksanderWWW) [reply mmdet/core/hook/neptunelogger_hook.py]: How do you mean? This class doesn't have an attribute like that. Plus numpy is listed in global requirements already (build requirements and runtime requirements I think)

Comment (kshitij12345) [reply mmdet/core/hook/neptunelogger_hook.py]: More context here : https://builtin.com/data-science/numpy-random-seed

Comment (AleksanderWWW) [reply mmdet/core/hook/neptunelogger_hook.py]: Got it. Good call. So shall I change it to sth like:
```# Set seed so that same validation set is logged each time.
    rng = np.random.default_rng(42)
    rng.shuffle(eval_image_indices)
    self.eval_image_indices = eval_image_indices[:self.num_eval_predictions]
```
?

File: mmdet/core/hook/neptunelogger_hook.py (hunk 9 above)
Comment (kshitij12345) [reply mmdet/core/hook/neptunelogger_hook.py]: I think this makes sense. In case of lightning we overwrite the `init_kwargs` if we have received `api_token` or `project` as an argument

https://github.com/Lightning-AI/lightning/blob/65eb3e7c456dc572d909362157ee2d29e99e2276/src/lightning/pytorch/loggers/neptune.py#L289-L296

File: mmdet/core/hook/neptunelogger_hook.py (hunk 11 above)
Comment (kshitij12345) [reply mmdet/core/hook/neptunelogger_hook.py]: I think this can we tweaked to the following so that we don't need to recompute the eval_image_indices. Wdyt?

```python
if not self.eval_image_indices:
  rng = np.random.default_rng(42)
  rng.shuffle(eval_image_indices)
  self.eval_image_indices = eval_image_indices[:self.num_eval_predictions]
```

File: configs/mask_rcnn/neptune_config.py (hunk 5 above)
Comment (kshitij12345) [reply configs/mask_rcnn/neptune_config.py]: I see.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+            labels = data_ann['labels']
+            # masks = data_ann.get('masks', None)
+
+            # Get dict of bounding boxes to be logged.
+            assert len(bboxes) == len(labels)
+            im = imshow_det_bboxes(
+                image, bboxes, labels, class_names=CLASSES, show=False)
+            im = im / im.max()
+
+            # TODO Add masks
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: stray comment?

File: mmdet/core/hook/neptunelogger_hook.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+                        ext='pth',
+                        final=False,
+                        mode='epoch') -> None:
+        assert mode in ['epoch', 'iter']
+
+        if self.ckpt_hook is None:
+            return
+
+        if final:
+            file_name = 'latest'
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: For final checkpoint, does it make sense to call it final rather than latest?

File: mmdet/core/hook/neptunelogger_hook.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+    @master_only
+    def after_run(self, runner) -> None:
+        """If enabled by the user, logs final model checkpoint, syncs and stops
+        the Neptune run."""
+        if self.log_model:
+            self._log_checkpoint(runner, final=True)
+
+        runner.logger.info('Syncing with Neptune.ai')
+        self._run.sync()
+        self._run.stop()
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py:333]: Does stop automatically call `sync`? In that case, we won't need the `sync` before.

**Comment by kshitij12345** [comment]
Also, looking at the failures, it seems that we shouldn't be importing neptune at file level. wandb does it within the logger class.

File: mmdet/core/hook/neptunelogger_hook.py [hunk 15]
```diff
@@ -153,12 +157,13 @@ def _add_ground_truth(self, runner):
...
         # Select the images to be logged.
-        eval_image_indices = np.arange(len(self.val_dataset))
-        # Set seed so that same validation set is logged each time.
-        rng = np.random.default_rng(42)
-        rng.shuffle(eval_image_indices)
-        self.eval_image_indices = eval_image_indices[:self.
-                                                     num_eval_predictions]
+        if not self.eval_image_indices:
+            eval_image_indices = np.arange(len(self.val_dataset))
+            # Set seed so that same validation set is logged each time.
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: I think we can remove this comment as we would be computing indices only once (so this maybe confusing).

File: mmdet/core/hook/neptunelogger_hook.py [hunk 16]
```diff
@@ -117,11 +117,11 @@ def __init__(self,
         self.num_eval_predictions = num_eval_predictions
         self.log_eval_predictions: bool = (num_eval_predictions > 0)
 
-        self.ckpt_hook: CheckpointHook | None = None
-        self.ckpt_interval: int | None = None
+        self.ckpt_hook: Union[CheckpointHook, None] = None
```
Comment (kshitij12345) [inline mmdet/core/hook/neptunelogger_hook.py]: I think the correct type would be `Optional[CheckpointHook]`.

Also applies to `ckpt_interval`, `eval_hook` and `eval_interval`.

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thanks @AleksanderWWW 

(Though, the lint errors look real)
